FHI-AIMS

Marcello-Sega opened the pull request:
Changes made in this Pull Request:

 - introduces support for FHI-AIMS coordinates files

PR Checklist
------------
 - [X] Tests?
 - [X] Docs?
 - [x] CHANGELOG updated?
 - [ ] Issue raised/referenced?

Marcello-Sega commented:
Hi, 

I'm working with FHI-AIMS since a while and jotted down a reader/writer for its simple, xyz-like input files. Maybe you find it useful.

Cheers

IAlibay commented:
Thanks for this awesome contribution @Marcello-Sega 
Quick initial question (apologies if I've missed this somewhere in the docs you've provided), do FHI-AIMS coordinates files have a defined filed format specification that we can refer to? (something like this: http://ambermd.org/netcdf/nctraj.xhtml).

This would be really useful (especially in the review process) in helping us better understand the expected "behaviour" of the coordinate files.

Marcello-Sega commented:
@IAlibay thanks for looking into this. 

I think the only authoritative source is the manual (not public), I copy-paste the relevant part below. Otherwise, the home page of the project is https://aimsclub.fhi-berlin.mpg.de/

Open Babel implements a reader and a writer for the format, and its source code is here: 
https://github.com/openbabel/openbabel/blob/master/src/formats/fhiaimsformat.cpp

If you think it will be enough, I could just add to the comments the reference to the webpage and a stripped-down version of the manual section:

- Tag: **`atom`** 
    Usage: `atom x y z species_name`
    Purpose: Specifies the initial location and type of an atom.
    `x`, `y`, `z` are real numbers (in Å) which specify the atomic position. `species_name` is a 
    string descriptor which names the element on this atomic position; it must match with one of 
    the species descriptions given in control.in.
- Tag: **`atom_frac`** (geometry.in)
    Usage: `atom_frac n1 n2 n3 species_name`
    Purpose: Specifies the initial location and type of an atom in fractional coordinates.
    `ni` is a real multiple of lattice vector `i`. species_name is a string descriptor which names the 
    element on this atomic position; it must match with one of the species descriptions given in 
    control.in.
    Fractional coordinates are only meaningful in periodic calculations.
- Tag: **`lattice_vector`** (geometry.in)
    Usage: `lattice_vector x y z`
    Purpose: Specifies one lattice vector for periodic boundary conditions.
    `x`,`y`, `z` are real numbers (in Å) which specify the direction and length of a unit cell vector.
    If up to three lattice vectors are specified, FHI-aims automatically assumes periodic boundary 
    conditions in those directions.

richardjgowers commented on package/MDAnalysis/coordinates/FHIAIMS.py:
```diff
@@ -0,0 +1,261 @@
+
+    def _read_first_frame(self):
+        with util.openany(self.filename, 'rt') as fhiaimsfile:
+            lines = fhiaimsfile.readlines()
```
I would try not to use `readlines()`.  If you accidentally pass in some multi TB file it will try and parse the entire file then fail without any error message.  Better would be to work through the file line by line, taking what you need and giving a coherent error message when something awry happens.

Marcello-Sega replied:
fixed

richardjgowers commented on package/MDAnalysis/coordinates/FHIAIMS.py:
```diff
@@ -0,0 +1,261 @@
+            self.ts._unitcell[:] = box
+            ts._pos = positions.copy()
+            # No velocities in FHIAIMS geometry files
+            velocities = np.zeros((n_atoms, 3), dtype=np.float32)
```
if there aren't velocities just don't make the array

Marcello-Sega replied:
fixed

richardjgowers commented on package/MDAnalysis/coordinates/FHIAIMS.py:
```diff
@@ -0,0 +1,261 @@
+Classes
+-------
+
+.. autoclass:: Timestep
```
this module will need a `.rst` stub to get documentation generated

Marcello-Sega replied:
fixed

orbeckst commented:
Following up on https://github.com/MDAnalysis/mdanalysis/pull/2705#issuecomment-636445719 : add the file format definition to the docs. Note that if the original docs are not public then you cannot copy and paste as you (and we) would be violating copyright. You need to rewrite the specs in your own words.

orbeckst commented on testsuite/MDAnalysisTests/datafiles.py:
```diff
@@ -187,7 +187,8 @@
     "PDB_CRYOEM_BOX", #Issue 2599
-    "PDB_CHECK_RIGHTHAND_PA" # for testing right handedness of principal_axes
+    "PDB_CHECK_RIGHTHAND_PA", # for testing right handedness of principal_axes
+    "FHIAIMS" # to test FHIAIMS coordinate files
```
add a comma to make it easier for the next addition
```suggestion
    "FHIAIMS", # to test FHIAIMS coordinate files
```

Marcello-Sega replied:
done

orbeckst commented on package/MDAnalysis/topology/FHIAIMSParser.py:
```diff
@@ -0,0 +1,108 @@
+
+Reads an FHI-AIMS .in file and pulls the atom information from it.  Because
+an FHI-AIMS input file only has atom name information, all information about residues
+and segments won't be populated.
```
Add description of file format (or link to the description in coordinates if the same)

Marcello-Sega replied:
added the link to the coordinates

orbeckst commented on package/MDAnalysis/coordinates/FHIAIMS.py:
```diff
@@ -0,0 +1,261 @@
+  version of the (3,3) triclinic vector representation of the unit
+  cell.  
+
+.. _FHIAIMS: https://aimsclub.fhi-berlin.mpg.de/
```
Add definition of the file format. Documentation included in MDAnalysis also falls under GPL so if you copy and paste from other sources, tell us source and license of those docs. If you write it in your own words then there shouldn't be a problem.

Marcello-Sega replied:
added my own description

Marcello-Sega commented:
Thanks for the helpful feedback, it looks to me like all has been fixed now, could you just review it shortly again? Also, I didn't generate the docs, so I'm not sure if they're error-free.

Marcello-Sega commented:
I solved (hopefully all) issues with sphinx, the only failing build is because of the exceeded time limit:
 
https://travis-ci.com/github/MDAnalysis/mdanalysis/jobs/342195774#L2110

This looks to me like an issue on Travis. Would it be possible to restart that job, or should I trigger the complete build with a new commit? Thanks

orbeckst commented:
I restarted the job – fingers crossed it makes it this time...

volkerblum commented:
This is amazing. I have not tried it but very excited to see this.

Quick comments:

1) FHI-aims will treat any line starting with '#' as a comment - i.e., a line ' # atom ...' will be treated as a comment. Does the current parser account for this? Would be good, if simple to do. 

(NB: This isn't too problematic in practice. Commenting an 'atom' line would normally be a user action, while the present use case is likely for files written by the code.)

2) There is a document that has the essence of the file format, as far as needed. Originally, this is intended, e.g., for viewers or to aid in structure conversion to different formats. The document therefore does not detail all keywords. The more code-specific keywords are in the manual. No strong reason for this, only that maintaining documentation in too many different places can become complicated. 

This document is in LaTeX format (written by myself) and this should certainly be publicly available somewhere. (Err, actually that is what I intended when I first put it together.) What's a good place for this purpose ... ?

[I also noticed that I need to add the 'velocity' definition to that document, for the present use.]

Best wishes & thank you!!
Volker

orbeckst commented:
@volkerblum thanks for the comments!

> This document is in LaTeX format (written by myself) and this should certainly be publicly available somewhere. (Err, actually that is what I intended when I first put it together.) What's a good place for this purpose ... ?

Ideally, put such docs on [figshare](https://figshare.com/) or [zenodo](https://zenodo.org/) – then it will get a DOI, be properly archived for posterity, and will have a clear license. This also means that we can link to a stable URL as a reference.

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_fhiaims.py:
```diff
@@ -0,0 +1,54 @@
+    return mda.Universe(FHIAIMS)
+
+
+def test_fhiaims_positions(FHIAIMS_U):
```
The tests only test the reader and not the writer.

Marcello-Sega replied:
Thanks for the feedback. I think I got a better grip on how the tests are handled (but it's not clear to me if there's a recommended way: I saw them organized using different approaches). 

Now I extended the coverage, including the writer. I fixed also string formatting, and comment lines are now handled as suggested by @volkerblum.

Please have a look at the last push

Marcello-Sega commented on testsuite/MDAnalysisTests/coordinates/test_fhiaims.py:
```diff
@@ -23,32 +23,159 @@
-    ts = FHIAIMS_U.trajectory[0]
-    assert_almost_equal(ts.dimensions,
-                        np.asarray([18.600000000000, 18.600000000000,  55.800000000000, 90., 90., 90.], dtype=np.float32))
+    def test_no_container(self, tmpdir, ref):
+        with tmpdir.as_cwd():
+            # load a configuration
+            u = mda.Universe(ref.filename)
+            # set up an FHIAIMS writer and write a frame
+            ref.writer('foo').write(u.atoms)
+            # extension should be added automatically
+            ref.reader('foo.in')
```
@orbeckst I'm not quite sure if this is testing what was meant to be tested.

orbeckst replied:
I don't think you need `test_no_container()` as this already exists in `base.BaseWriterTest` https://github.com/MDAnalysis/mdanalysis/blob/0911ff3a0379356ee603ff4aa21eec3b990f9958/testsuite/MDAnalysisTests/coordinates/base.py#L593-L599

Hopefully, just using `BaseWriterTest` as you're doing should cover everything. You can always look at the coverage report (the diff part) to see what was not tested.

Marcello-Sega replied:
OK. I need to set `container_format=True` then, I assume. What does actually define a format to be a container?

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_fhiaims.py:
```diff
@@ -23,32 +23,159 @@
+        assert_equal(ref.time, universe.trajectory.time,
+                     "wrong time of the frame")
+
+    def test_bad_input1(self, tmpdir):
```
Something like this is good.

You can actually create the file as a string buffer with `StringIO` and use it directly as show in https://github.com/MDAnalysis/mdanalysis/blob/0911ff3a0379356ee603ff4aa21eec3b990f9958/testsuite/MDAnalysisTests/coordinates/test_gro.py#L432-L440

I would make it a fixture and then use it somehow like this:
```python
from six import StringIO

@pytest.fixture(scope="module")
def bad_input_missing_periodicity():
    buffer = """
lattice_vector 1.0 0.0 0.0
lattice_vector 0.0 1.0 0.0
atom 0.1 0.1 0.1 H
"""
   return StringIO(buffer)

...

   def test_bad_input1(self, bad_input_missing_periodicity):
        with pytest.raises(ValueError, match="Found partial periodicity"):
             u = mda.Universe(bad_input_missing_periodicity, format="FHIAIMS")
```

Also note the use of `pytest.raises` to check for the correct exception and `match` to check for the error message itself.

Marcello-Sega replied:
Excellent, just made it class-wide

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_fhiaims.py:
```diff
@@ -23,32 +23,159 @@
+            fd.write('lattice_vector 1.0 0.0 0.0\n')
+            fd.write('lattice_vector 0.0 1.0 0.0\n')
+            fd.write('atom 0.1 0.1 0.1 H')
+        try:
```
rewrite with `pytest.raises`

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_fhiaims.py:
```diff
@@ -23,32 +23,159 @@
+                    "FHIAIMSReader did not detect a missing lattice_vector")
+
+    def test_bad_input2(self, tmpdir):
+        fname = str(tmpdir.mkdir("FHIAIMSReader").join('bad_input2.in'))
```
Make it a StringIO fixture `bad_input_illegal_line` or something like that (words are better than numbers).

Marcello-Sega replied:
done

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_fhiaims.py:
```diff
@@ -23,32 +23,159 @@
+        fname = str(tmpdir.mkdir("FHIAIMSReader").join('bad_input2.in'))
+        with open(fname, 'w') as fd:
+            fd.write('garbage')
+        try:
```
rewrite with `pytest.raises`

Marcello-Sega replied:
see above

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_fhiaims.py:
```diff
@@ -23,32 +23,159 @@
+                    "FHIAIMSReader did not detect a non-conforming line")
+
 
+class TestFHIAIMSWriter(BaseWriterTest):
```
Yes – thanks for adding the test!

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_fhiaims.py:
```diff
@@ -23,32 +23,159 @@
-    ts = FHIAIMS_U.trajectory[0]
-    assert_almost_equal(ts.dimensions,
-                        np.asarray([18.600000000000, 18.600000000000,  55.800000000000, 90., 90., 90.], dtype=np.float32))
+    def test_no_container(self, tmpdir, ref):
```
Can probably be removed as it is in `BaseWriterTest`

orbeckst replied:
Also, the current test does not `assert` anything.

Marcello-Sega replied:
removed

volkerblum commented:
Ok, so after 15 1/2 years a description of the basic FHI-aims geometry.in file format is now online and citeable with a DOI:

Blum, Volker (2020): FHI-aims File Format Description: geometry.in. figshare. Online resource. https://doi.org/10.6084/m9.figshare.12413477.v1

Note that the part of the format described here is actually quite general and practical and can be read by several utilities outside of FHI-aims. The description focuses on those parts of the format that are generic and deliberately omits technical keywords related to operating the FHI-aims code itself.

Apologies in case I missed any other citeable instance of this or of a similar document. I wrote this one.

Marcello-Sega commented on testsuite/MDAnalysisTests/coordinates/test_fhiaims.py:
```diff
@@ -23,32 +23,159 @@
+
+    def test_n_frames(self, ref, universe):
+        assert_equal(ref.n_frames, universe.trajectory.n_frames,
+                     "wrong number of frames")
```
The format is single-frame only

orbeckst replied:
ok

Marcello-Sega commented on package/MDAnalysis/topology/FHIAIMSParser.py:
```diff
@@ -84,11 +84,13 @@ def parse(self, **kwargs):
         # FHIAIMS geometry files are only single frames
         names = []
         with openany(self.filename) as inf:
-            names, line  = [], inf.readline()
```
No particular reason, I will switch.

Marcello-Sega commented on package/MDAnalysis/coordinates/FHIAIMS.py:
```diff
@@ -142,9 +142,11 @@ def _read_first_frame(self):
                     relative.append('atom_frac' in line)
                 elif 'lattice' in line:
                     lattice_vectors.append(list(np.asarray(line.split()[1:])))
+                elif line[0] == '#':
```
ditto

Marcello-Sega commented on testsuite/MDAnalysisTests/coordinates/test_fhiaims.py:
```diff
@@ -23,32 +23,159 @@
+    def universe2(self):
+        return mda.Universe(FHIAIMS)
+
+ #   def setUp(self):
```
ok

Marcello-Sega commented:
Most welcome @orbeckst, MDAnalysis is a great tool and I'm using it on a daily basis. Happy to contribute a little bit. Actually, today I showed this thread to my guys as an example of how to conduct code review.

Just pushed: I implemented the requested changes, updated the link to the new docs (thanks @volkerblum). The code now handles also the velocity tag.

orbeckst commented on package/MDAnalysis/coordinates/FHIAIMS.py:
```diff
@@ -0,0 +1,322 @@
+            oldline=''
+            for line in fhiaimsfile:
+                line = line.strip()    # perhaps not needed
+                if any([line.startswith(tag) for tag in skip_tags] ):
```
remove space
```suggestion
                if any([line.startswith(tag) for tag in skip_tags]):
```

orbeckst replied:
I would move this `if` statement to after the ones checking for `atom` and `velocity`: I assume there will be more atoms than skip lines so you save many evaluations of the skip-if.

orbeckst commented on package/MDAnalysis/coordinates/FHIAIMS.py:
```diff
@@ -0,0 +1,322 @@
+            for line in fhiaimsfile:
+                line = line.strip()    # perhaps not needed
+                if any([line.startswith(tag) for tag in skip_tags] ):
+                    oldline=line
```
pep8
```suggestion
                    oldline = line
```

orbeckst commented on package/MDAnalysis/coordinates/FHIAIMS.py:
```diff
@@ -0,0 +1,322 @@
+            skip_tags = ["#","initial_moment"]
+            oldline=''
+            for line in fhiaimsfile:
+                line = line.strip()    # perhaps not needed
```
Is needed because the [format doc](https://figshare.com/articles/FHI-aims_File_Format_Description_geometry_in/12413477/1) said "white space separated":
> In geometry.in, whitespace separates keywords and their values.  Empty lines are ignored andcomment lines can be inserted starting with the ’#’ character. 

Remove the comment

orbeckst commented on package/MDAnalysis/coordinates/FHIAIMS.py:
```diff
@@ -0,0 +1,322 @@
+                if line.startswith("atom"):
+                    positions.append(
+                        list(np.asarray(line.split()[1:-1], np.float32)))
+                    relative.append('atom_frac' in line)
```
neat

orbeckst commented on package/MDAnalysis/coordinates/FHIAIMS.py:
```diff
@@ -0,0 +1,322 @@
+                    continue
+                if line.startswith("atom"):
+                    positions.append(
+                        list(np.asarray(line.split()[1:-1], np.float32)))
```
Can't you do the conversion at the very end, i.e.,
```suggestion
                        line.split()[1:-1])
```
and then later
```python
positions = np.array(positions, dtype=np.float32)
```
Should improve performance because you don't go back and forth between arrays and lists and numpy can cast everything in one go.

orbeckst replied:
You're already doing 
```python
positions = np.asarray(positions, dtype=np.float32)
``` 
so definitely don't use np.array here.

Marcello-Sega replied:
done

orbeckst commented on package/MDAnalysis/coordinates/FHIAIMS.py:
```diff
@@ -0,0 +1,322 @@
+                        raise ValueError(
+                            'Non-conforming line (velocity must follow atom): ({0})in FHI-AIMS input file {0}'.format(line, self.filename))
+                    velocities.append(
+                        list(np.asarray(line.split()[1:], np.float32)))
```
same as above – keep positions a list of lists and convert to float32 array at end?

Marcello-Sega replied:
yep

orbeckst commented on package/MDAnalysis/coordinates/FHIAIMS.py:
```diff
@@ -0,0 +1,322 @@
+            positions = np.asarray(positions, dtype=np.float32)
+            velocities = np.asarray(velocities, dtype=np.float32)
+            relative = np.asarray(relative)
+            if len(velocities) not in [0,len(positions)]:
```
pep8

orbeckst commented on package/MDAnalysis/coordinates/FHIAIMS.py:
```diff
@@ -0,0 +1,322 @@
+                    'Found partial periodicity in FHI-AIMS file. This cannot be handled by MDAnalysis.')
+            if len(lattice_vectors) == 0 and any(relative):
+                raise ValueError(
+                    'Found relative coordinates in FHI-AIMS file withut lattice info.')
```
```suggestion
                    'Found relative coordinates in FHI-AIMS file without lattice info.')
```

orbeckst commented on package/MDAnalysis/coordinates/FHIAIMS.py:
```diff
@@ -0,0 +1,322 @@
+            self.n_atoms = n_atoms = len(positions)
+            self.ts = ts = self._Timestep(n_atoms, **self._ts_kwargs)
+
+            if len(lattice_vectors)>0:
```
pep8

orbeckst commented on package/MDAnalysis/coordinates/FHIAIMS.py:
```diff
@@ -0,0 +1,322 @@
+
+            if len(lattice_vectors)>0:
+                box = np.asarray(lattice_vectors, dtype=np.float32)
+                positions[np.where(relative)] = np.matmul(
+                    positions[np.where(relative)], box)
```
I think this can be written simply as
```suggestion
                positions[relative] = np.matmul(positions[relative], box)
```
I really liked the idea of the `relative` bool array.

orbeckst replied:
Is the order of the multiplication correct? I assume some broadcasting is needed. 

Do you have a test case for fractional?

Marcello-Sega replied:
I thought so.  Added a test anyway, and passes.

orbeckst commented on package/MDAnalysis/coordinates/FHIAIMS.py:
```diff
@@ -0,0 +1,322 @@
+    """
+
+    format = ['IN', 'FHIAIMS']
+    units = {'time': None, 'length': 'Angstrom', 'velocity': 'Angstrom/ps'}
```
time ought to be ps then

Marcello-Sega replied:
right :)

orbeckst commented on package/MDAnalysis/coordinates/FHIAIMS.py:
```diff
@@ -0,0 +1,322 @@
+
+
+class FHIAIMSWriter(base.WriterBase):
+    """FHIAIMS Writer that conforms to the Trajectory API.
```
see comment on docstring for the Reader

Marcello-Sega replied:
I guess it's clean now.

orbeckst commented on package/MDAnalysis/coordinates/FHIAIMS.py:
```diff
@@ -0,0 +1,322 @@
+
+    #: format strings for the FHIAIMS file (all include newline)
+    fmt = {
+        # coordinates output format, see http://chembytes.wikidot.com/g-fhiaimsfile
```
and link to the official format doc?

Marcello-Sega replied:
fixed

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_fhiaims.py:
```diff
@@ -0,0 +1,208 @@
+    changing_dimensions = False
+
+
+class TestFHIAIMSReader(object):
```
Is there a test for testing the`atom_frac` reading (for all and some)?

Marcello-Sega replied:
there now

orbeckst commented on testsuite/MDAnalysisTests/coordinates/test_fhiaims.py:
```diff
@@ -0,0 +1,208 @@
+
+    prec = 6
+
+    @pytest.fixture(scope='class')
```
nice!

Marcello-Sega commented on package/MDAnalysis/coordinates/FHIAIMS.py:
```diff
@@ -0,0 +1,322 @@
+
+    lattice_vector x  y  z
+
+where x,y, and z are expressed in Angstrom.
```
` git grep "units.*ngstr"`  shows an overwhelming prevalence of Angstrom, let's leave it like that, it will be easier to make a bulk change later on.

Marcello-Sega replied:
fixed with [ångström]

Marcello-Sega commented on package/MDAnalysis/coordinates/FHIAIMS.py:
```diff
@@ -0,0 +1,322 @@
+
+    """
+    format = ['IN', 'FHIAIMS']
+    units = {'time': None, 'length': 'Angstrom', 'velocity': None}
```
fixed

Marcello-Sega commented on package/MDAnalysis/coordinates/FHIAIMS.py:
```diff
@@ -0,0 +1,322 @@
+    def _read_first_frame(self):
+        with util.openany(self.filename, 'rt') as fhiaimsfile:
+            relative, positions, velocities, lattice_vectors = [], [], [], []
+            skip_tags = ["#","initial_moment"]
```
I will autopep8 the whole thing again.

Marcello-Sega commented on package/MDAnalysis/coordinates/FHIAIMS.py:
```diff
@@ -0,0 +1,322 @@
+        with util.openany(self.filename, 'rt') as fhiaimsfile:
+            relative, positions, velocities, lattice_vectors = [], [], [], []
+            skip_tags = ["#","initial_moment"]
+            oldline=''
```
ditto

Marcello-Sega commented on package/MDAnalysis/coordinates/FHIAIMS.py:
```diff
@@ -0,0 +1,322 @@
+attribute, which is a dictionary of different strings for writing
+lines in ``.in`` files.  These are as follows:
+
+``xyz``
+  An atom line without velocities.  Requires that the 'name' and
+  'pos' keys be supplied.  E.g.::
+
+     fmt['xyz'].format(pos=(0.0, 1.0, 2.0), name='O')
+
+``box_triclinic``
+  The (optional) initial lines of the file which gives box dimensions.
+  Requires the box keyword, as a length 9 vector.  This is a flattened
+  version of the (3,3) triclinic vector representation of the unit
+  cell.
```
Tried before but got this error 
```
sphinx.errors.SphinxWarning: /Users/sega/mdanalysis/package/MDAnalysis/analysis/base.py:docstring of MDAnalysis.analysis.base.AnalysisFromFunction:28:Field list ends without a blank line; unexpected unindent.

Warning, treated as error:
/Users/sega/mdanalysis/package/MDAnalysis/analysis/base.py:docstring of MDAnalysis.analysis.base.AnalysisFromFunction:28:Field list ends without a blank line; unexpected unindent.
```
and didn't want to change it. I will include a fix.

Marcello-Sega replied:
Sphinx (before: 2.x, now updated to 3.0.4) throws tens of similar errors, I guess the problem is somewhere else, but like that, I can't build the docs, sorry (this is a nice reminder to myself of why I stopped using sphinx long ago...)

orbeckst commented on package/MDAnalysis/coordinates/FHIAIMS.py:
```diff
@@ -0,0 +1,333 @@
+                tri_dims = obj.universe.coord.triclinic_dimensions
+            except AttributeError:  # for Timestep
+                tri_dims = obj.triclinic_dimensions
+            # full output
+            try:
+                box = tri_dims.flatten()
+                output_fhiaims.write(self.fmt['box_triclinic'].format(box=box))
+            except (NameError, UnboundLocalError):
+                pass  # no box information
```
I don't get the logic here. Can you explain? My thinking is:

Presumably, you don't want to write `lattice_vector` lines if there's no box. Under which circumstances do you get a `NameError` or a `UnboundLocalError` here ? You should always be able to get this information. Both are exceptions that I associate with unclear code paths. We should be more explicit here (even though in general I do like Python's "it's easier to ask for forgiveness than for permission" approach with exceptions.

There's also an additional consideration: We are not yet consistent in how we flag "no PBC" (see #2698 ): at the moment, `dimensions` can either be `None` (which we will adapt as the standard following #2698) or `[0,0,0,0,0,0]` — either way, `triclinic_vectors` will return a 3x3 array of zeros and that will write a non-sensical output line (I think). Luckily, using `triclinic_vectors` avoids the current issue with #2698 but one needs to test for zero-vectors.

 I suggest something like the following where we check that at least one element of the triclinic box vectors is non-zero. Perhaps a better test would check that at least one in each row is non-zero or that the volume is greater than zero (`ts.volume`).
```suggestion
           # full output
           if np.any(tri_dims != 0):      
                output_fhiaims.write(self.fmt['box_triclinic'].format(box=tri_dims.flatten()))
```

If you wanted to check the volume then I would change the preceding code to give you the `ts` object instead of `tri_dims` and then work with `ts.tri_dims` and `ts.volume`.

Marcello-Sega replied:
Totally right.  The `np.any(tri_dims != 0)` is fine.

Marcello-Sega commented on package/MDAnalysis/coordinates/FHIAIMS.py:
```diff
@@ -0,0 +1,333 @@
+            warnings.warn(
+                "Supplied AtomGroup was missing the following attributes: "
+                "{miss}. These will be written with default values. "
+                "Alternatively these can be supplied as keyword arguments."
```
Leftover of a copy&paste from `coordinates/GRO.py`, please have a look at its `write` method, has the same issue.

Marcello-Sega commented on testsuite/MDAnalysisTests/coordinates/test_fhiaims.py:
```diff
@@ -0,0 +1,226 @@
+    changing_dimensions = False
+
+
+class TestFHIAIMSReader(object):
```
Done

Marcello-Sega commented:
Should be fine now. The only thing I wasn't able to check is how the docs are rendered (couldn't make sphinx work properly). If you could have a look, that would be great.

orbeckst commented on package/MDAnalysis/coordinates/FHIAIMS.py:
```diff
@@ -0,0 +1,333 @@
+                    positions[relative], lattice_vectors)
+                self.ts._unitcell[:] = lattice_vectors
+
+            ts._pos = positions.copy()
```
I don't think we need to make a copy here because the `_pos` setter fills the array anyway. Avoiding the copy should improve performance. — @richardjgowers ?

orbeckst replied:
Er, sorry, `ts._pos` is the actual array. If we used
```
ts.positions = positions
```
as we do in, for instance, in the XYZReader, then we fill the array. The latter is cleaner and makes sure that the dtype stays. We also don't need the asarray casting above, that's done with the assignment to a np array of a given dtype.

orbeckst replied:
Tried `ts.positions = positions` and it works. Will push my changes.

Marcello-Sega replied:
> I'll look at the docs and make any other changes.

Thanks!

orbeckst commented:
I'll look at the docs and make any other changes.

orbeckst commented:
I am adding a test for the parser and then I will rewrite the history and force-push to your branch. No need to do anything else from your end. We'll then just wait for the CI to come back green.

orbeckst commented:
Once the CI shows no errors, this can be *merged*.

orbeckst commented:
@Marcello-Sega thank you for your first contribution! Congratulations 🎉 ! It took a little longer than perhaps initially thought but you know have code in MDAnalysis that passes all our quality standards and fits the library style. I hope we see more from you!

If you want GitHub to list your contribution, add your fz email address to your GitHub account. 

I am also sending you an invite for the mdanalysis org as a contributor.

Marcello-Sega commented:
> @Marcello-Sega thank you for your first contribution! Congratulations 🎉 ! It took a little longer than perhaps initially thought but you now have code in MDAnalysis that passes all our quality standards and fits the library style. I hope we see more from you!
> 
> If you want GitHub to list your contribution, add your fz email address to your GitHub account.
> 
> I am also sending you an invite for the mdanalysis org as a contributor.

Nice, thanks! yep, it would be nice to discuss some ideas in the future...

orbeckst commented:
All tests passed, test coverage is almost complete – well done!! Congratulations, @Marcello-Sega !

richardjgowers commented:
@Marcello-Sega thanks for the patch!  If you've got more ideas then feel free to raise issues to start discussion

volkerblum commented:
Thank you so much, everyone, especially @Marcello-Sega , this is great!

One more question from me - do you perhaps have a good tutorial-like example (a minimal one would be sufficient) that we could include with FHI-aims to show how MDanalysis can be used, assuming one has output from FHI-aims?

(Knowing about MDanalysis would be immensely helpful for other users of the FHI-aims, I am sure.)

Again, thanks!

Marcello-Sega commented:
@volkerblum: a quick answer is that probably an adapted version of the simple example from the [README](https://github.com/MDAnalysis/mdanalysis/blob/develop/README.rst) could be use for a single frame case, or to show how to convert a series of files to/from a trajectory in another format (say, DCD). 

Personally, I'm using this in combination with [pytim](https://github.com/Marcello-Sega/pytim/blob/master/README.md) for surface analysis, maybe this can also be interesting. I could, however, write some more meaningful examples if you can provide a list of 3 or 4 cases that in your opinion represent the typical workflows (including which file types are involved: I guess that [cube](http://paulbourke.net/dataformats/cube/) will be one of them:  this could be handled also easily with MDAnalysis).  What do you think?

I hope this discussion is not off-topic here. If @richardjgowers or @orbeckst suggest so we could move it somewhere else.

volkerblum commented:
Ah, I see. Good question.

Regarding the link to the README: I didn't see FHI-aims under "6.1.4. Supported coordinate formats" yet in the documentation. 

Should it be? (https://www.mdanalysis.org/docs/documentation_pages/coordinates/init.html#id1) [Sorry, I'm pretty sure I'm missing something simple. I guess that question, at least, is hopefully still on-topic for this thread.]

pytim sounds very useful, thank you for the pointer. 

Regarding typical workflows - my limited experience is that in MD, there are as many typical modes of analysis as there are scientists. 

That being said, even something simple like trajectory averaging (over the past n steps, 7.5.3 in documentation) could be a good enough example. 

pytim itself (+ MDAnalysis) could also be a good example.

Any of these?

Thanks again!

orbeckst commented:
@volkerblum @Marcello-Sega feel free to continue the conversation here. We're very happy to add FHI-AIMS to MDA (just in time for 1.0...).

If you have tutorials, we're also happy to list them on our website https://www.mdanalysis.org/pages/learning_MDAnalysis/

orbeckst commented:
> Regarding the link to the README: I didn't see FHI-aims under "6.1.4. Supported coordinate formats" yet in the documentation.

@volkerblum It's in the development docs (slightly different url) https://www.mdanalysis.org/mdanalysis/documentation_pages/coordinates/init.html#id2 ; once the 1.0 release is out it will be in the docs that you linked, too.
